cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp

```diff
@@ -0,0 +1,272 @@
+ *  limitations under the License.
+ ******************************************************************************/
+#include "CoordinateConversionController.h"
+
+#include "CoordinateConversionResult.h"
+#include "CoordinateOptionDefaults.h"
+#include "GeoViews.h"
+
+// Qt headers
+#include <QtGlobal>
+
+// ArcGISRuntime headers
+#include <CoordinateFormatter.h>
+
+namespace Esri
+{
+namespace ArcGISRuntime
+{
+namespace Toolkit
+{
+
+namespace
+{
+constexpr double DEFAULT_ZOOMTO_DISTANCE = 1500.0;
+}
+
+CoordinateConversionController::CoordinateConversionController(QObject* parent):
+  QObject(parent),
+  m_zoomToDistance(DEFAULT_ZOOMTO_DISTANCE),
+  m_coordinateFormats(new GenericListModel(&CoordinateConversionOption::staticMetaObject, this)),
+  m_conversionResults(new GenericListModel(&CoordinateConversionResult::staticMetaObject, this)),
+  m_geoView(nullptr),
+  m_inPickingMode(false)
+{
+  {
+    // Some default coordinate conversion formats to set us up with.
+    m_coordinateFormats->setDisplayPropertyName("name");
+    m_coordinateFormats->append(QList<QObject*>()
+      << createDecimalDegrees()
+      << createDegreesDecimalMinutes()
+      << createDegreesMinutesSeconds()
+      << createMgrs()
+      << createUsng()
+      << createUtm()
+      << createGars()
+    );
+  }
+
+  {
+    m_conversionResults->setDisplayPropertyName("name");
+    connect(m_conversionResults, &QAbstractItemModel::rowsInserted, this,
+      [this](const QModelIndex& /*parent*/, int first, int last)
+      {
+        // Every result owned by the controller needs to listen to the
+        // currentPointChanged signal to update its notation.
+        for (int i = first; i <= last; ++i)
+        {
+          auto index = m_conversionResults->index(i);
+          auto result = m_conversionResults->element<CoordinateConversionResult>(index);
+          if (result)
+            connect(this, QOverload<const Point&>::of(&CoordinateConversionController::currentPointChanged),
+                    result, QOverload<const Point&>::of(&CoordinateConversionResult::updateCoordinatePoint));
+          else
+            qWarning("Empty CoordinateConversionResult in the CoordinateConversionController results model at index %d", i);
+        }
+      }
+     );
+  }
+}
+
+CoordinateConversionController::~CoordinateConversionController()
+{
+}
+
+QObject* CoordinateConversionController::geoView() const
+{
+  return m_geoView;
+}
+
+void CoordinateConversionController::setGeoView(QObject* geoView)
+{
+  if (geoView == m_geoView)
+    return;
+
+  if (m_geoView)
+    disconnect(this, nullptr, m_geoView, nullptr);
+
+  m_geoView = geoView;
+
+  if (auto sceneView = qobject_cast<SceneViewToolkit*>(m_geoView))
+  {
+    connect(sceneView, &SceneViewToolkit::mouseClicked, this,
+            [sceneView, this](QMouseEvent& event)
+    {
+      if (m_inPickingMode)
+      {
+        setCurrentPoint(sceneView->screenToBaseSurface(event.x(), event.y()));
```

Comment:
this won't take account of buildings etc. for the scene view. Probably better to use this than the async call though...

Reply:
Agreed. 
This was lifted from the old tool -  if you think this should be changed then this would be the time.